### PR TITLE
fixes #520

### LINKED ISSUE #520: Context.userId containing ResourceType and ResourceId causes problems
Hi Folks, 

With the `patient-view` hook, there are three context variables: `userId`, `patientId`, `encounterId`. 
Of these `patientId` and `encounterId` are just IDs. 
However, `userId` is unique - it has the ResourceType and the ID separated by "/" in it.
This causes problems. 

For example, with `patientId` I can formulate a query like:
`patient?_id={{context.patientId}}`
This is perfectly legitimate, and is sometimes necessary to convert a read into a search to add additional search parameters such as `_include` etc.

However, if I try the same with `userId`, then say the query is as below and the ID is 123
`practitioner?_id={{context.userId}}`
This leads to two problems:
1. The userId may or may not be a practitioner resource. If it is not, and say it's a Patient, there is no way to tell not to try this, and we have failed - or worse, if there *happened to be* a practitioner resource with the same ID as the patient ID, what does that imply?
2. When translated, this query becomes
`practitioner?_id=Practitioner/123`
and this just fails because of the repetition of the resource type!

Can we do something to address this? 
/matt

### PR DESCRIPTION
**Pull request by isaacvetter**
And jira: https://jira.hl7.org/browse/FHIR-25761

Fixes #520

### CODE REVIEW

**Review by brynrhodes — APPROVED** [review]
Agree with the approach on the grounds that it effectively addresses the problem of prefetch template writing, does not impact existing hook definitions, and does not require hook definitions to change to take advantage of it.

File: docs/specification/current.md [hunk 1]
```diff
@@ -226,11 +226,43 @@ It is the CDS Service's responsibility to check prefetched data against its temp
...
+
+A prefetch template may include any of the following prefetch tokens:
+
+
+Token | Description
+---|---
+`{{userPractitionerId}}` | FHIR id of the Practitioner resource corresponding to the current user. 
+`{{userPractitionerRoleId}}`|FHIR id of the Practitioner resource corresponding to the current user. 
+`{{userPatientId}}`|FHIR id of the Practitioner resource corresponding to the current user. 
+`{{userRelatedPersonId}}`|FHIR id of the Practitioner resource corresponding to the current user.
```
Comment (dennispatterson) [inline docs/specification/current.md:261]: Update "Practitioner" to be the appropriate type
```suggestion
`{{userPractitionerRoleId}}`|FHIR id of the PractitionerRole resource corresponding to the current user. 
`{{userPatientId}}`|FHIR id of the Patient resource corresponding to the current user. 
`{{userRelatedPersonId}}`|FHIR id of the RelatedPerson resource corresponding to the current user. 
```

File: docs/specification/current.md [hunk 2]
```diff
@@ -226,11 +226,43 @@ It is the CDS Service's responsibility to check prefetched data against its temp
...
+  "prefetch": {
+    "user": "{{context.userId}}"
+  }
+}
+```
+or though a FHIR search:
+```
+{
+  "prefetch": {
+    "user": "PractitionerRole?_id={{userPractitionerId}}&_include=PractitionerRole:practitioner"
```
Comment (dennispatterson) [inline docs/specification/current.md:248]: Wouldn't this be userPractitionerRoleId?
```suggestion
    "user": "PractitionerRole?_id={{userPractitionerRoleId}}&_include=PractitionerRole:practitioner"
```

File: docs/specification/current.md [hunk 3]
```diff
@@ -226,11 +226,43 @@ It is the CDS Service's responsibility to check prefetched data against its temp
...
+
+Token | Description
+---|---
+`{{userPractitionerId}}` | FHIR id of the Practitioner resource corresponding to the current user. 
+`{{userPractitionerRoleId}}`|FHIR id of the Practitioner resource corresponding to the current user. 
+`{{userPatientId}}`|FHIR id of the Practitioner resource corresponding to the current user. 
+`{{userRelatedPersonId}}`|FHIR id of the Practitioner resource corresponding to the current user. 
+
+
+No single FHIR resource represents a user, rather Practitioner and PractitionerRole may be jointly used to represent a provider or other, and Patient or Person are used to represent a patient or their proxy. Hook definitions typically define a `context.userId` field and corresponding prefetch token.
```
Comment (dennispatterson) [inline docs/specification/current.md:264]: Did you mean RelatedPerson?
```suggestion
No single FHIR resource represents a user, rather Practitioner and PractitionerRole may be jointly used to represent a provider or other, and Patient or RelatedPerson are used to represent a patient or their proxy. Hook definitions typically define a `context.userId` field and corresponding prefetch token.
```

Comment (dennispatterson) [reply docs/specification/current.md:264]: What does "or other" mean?  Should this be removed or am I misunderstanding?
```suggestion
No single FHIR resource represents a user, rather Practitioner and PractitionerRole may be jointly used to represent a provider, and Patient or Person are used to represent a patient or their proxy. Hook definitions typically define a `context.userId` field and corresponding prefetch token.
```

File: docs/specification/current.md [hunk 4]
```diff
@@ -226,11 +226,43 @@ It is the CDS Service's responsibility to check prefetched data against its temp
 
 #### Prefetch tokens
 
-A prefetch token is a placeholder in a prefetch template that is replaced by a value from the hook's context to construct the FHIR URL used to request the prefetch data.
+A prefetch token is a placeholder in a prefetch template that is *_replaced by information from the hook's context_* to construct the FHIR URL used to request the prefetch data.
 
-Prefetch tokens MUST be delimited by `{{` and `}}`, and MUST contain only the qualified path to a hook context field.
+Prefetch tokens MUST be delimited by `{{` and `}}`, and MUST contain only the qualified path to a hook context field *_or one of the following user identifiers: `userPractitionerId`, 'userPractitioneRolerId', `userPatientId`, or `userRelatedPersonId`_*.
```
Comment (dennispatterson) [inline docs/specification/current.md:231]: minor typo
```suggestion
Prefetch tokens MUST be delimited by `{{` and `}}`, and MUST contain only the qualified path to a hook context field *_or one of the following user identifiers: `userPractitionerId`, 'userPractitioneRoleId', `userPatientId`, or `userRelatedPersonId`_*.
```

**Review by dennispatterson — CHANGES_REQUESTED** [review]

File: docs/specification/current.md (hunk 1 above)
Comment (isaacvetter) [reply docs/specification/current.md:261]: thank you, Dennis!

File: docs/specification/current.md [hunk 5]
```diff
@@ -226,11 +226,43 @@ It is the CDS Service's responsibility to check prefetched data against its temp
...
+
+Token | Description
+---|---
+`{{userPractitionerId}}` | FHIR id of the Practitioner resource corresponding to the current user. 
+`{{userPractitionerRoleId}}`|FHIR id of the PractitionerRole resource corresponding to the current user. 
+`{{userPatientId}}`|FHIR id of the Patient resource corresponding to the current user. 
+`{{userRelatedPersonId}}`|FHIR id of the RelatedPerson resource corresponding to the current user. 
+
+
+No single FHIR resource represents a user, rather Practitioner and PractitionerRole may be jointly used to represent a provider, and Patient or Person are used to represent a patient or their proxy. Hook definitions typically define a `context.userId` field and corresponding prefetch token.
```
Comment (isaacvetter) [inline docs/specification/current.md:264]:
```suggestion
No single FHIR resource represents a user, rather Practitioner and PractitionerRole may be jointly used to represent a provider, and Patient or RelatedPerson are used to represent a patient or their proxy. Hook definitions typically define a `context.userId` field and corresponding prefetch token.
```

**Review by dennispatterson — APPROVED** [review]

**Pull request merged**